sdks/agent-sdk/src/core/Agent.ts

```diff
@@ -124,52 +166,130 @@ export class Agent<ContentTypes> extends EventEmitter<
       this.#isListening = true;
       void this.emit("start");
 
-      const stream = await this.#client.conversations.streamAllMessages();
+      const stream =
+        await this.#client.conversations.streamAllMessages(options);
       for await (const message of stream) {
         // The "stop()" method sets "isListening"
         // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
         if (!this.#isListening) break;
         try {
           await this.#processMessage(message);
         } catch (error) {
-          this.#throwError(error);
+          await this.#runErrorChain(error, null);
         }
       }
     } catch (error) {
       this.#isListening = false;
-      this.#throwError(error);
+      await this.#runErrorChain(error, null);
     }
   }
 
   async #processMessage(message: DecodedMessage<ContentTypes>) {
-    const conversation = await this.#client.conversations.getConversationById(
-      message.conversationId,
-    );
+    let context: AgentContext<ContentTypes> | null = null;
+    try {
+      const conversation = await this.#client.conversations.getConversationById(
+        message.conversationId,
+      );
 
-    if (!conversation) {
-      this.#throwError(
-        new Error(
+      if (!conversation) {
+        throw new Error(
           `Failed to process message ID "${message.id}" for conversation ID "${message.conversationId}" because the conversation could not be found.`,
-        ),
-      );
-      return;
-    }
+        );
+      }
 
-    const context = new AgentContext(message, conversation, this.#client);
+      context = new AgentContext(message, conversation, this.#client);
+      await this.#runMiddlewareChain(context);
+    } catch (error) {
+      await this.#runErrorChain(error, context);
+    }
+  }
 
-    let middlewareIndex = 0;
-    const next = async () => {
-      if (middlewareIndex < this.#middleware.length) {
-        const currentMiddleware = this.#middleware[middlewareIndex++];
-        await currentMiddleware(context, next);
-      } else if (filter.notFromSelf(message, this.#client)) {
-        // Note: we are filtering the agent's own message to avoid
-        // infinite message loops when a "message" listener replies
+  async #runMiddlewareChain(context: AgentContext<ContentTypes>) {
+    const finalEmit = () => {
+      if (filter.notFromSelf(context.message, this.#client)) {
         void this.emit("message", context);
       }
     };
 
-    await next();
+    const chain = this.#middleware.reduceRight<Parameters<AgentMiddleware>[1]>(
+      (next, mw) => {
+        return async () => {
+          try {
+            await mw(context, next);
+          } catch (error) {
+            const resume = await this.#runErrorChain(error, context);
+            if (resume) {
+              await next();
+            }
+            // Chain is not resuming, error is being swallowed
+          }
+        };
+      },
+      finalEmit,
+    );
```

Comment:
can you explain this chain? is the intention to run all middleware?

Reply:
Yes, it chains them: A(B(C(D(E(...))))) (final emit). I am using `reduceRight` to start from `E(...)` which then gets wrapped into `D(E(...))`.